### PR TITLE
interfaces: honor static attributes when reloading conns

### PR DESCRIPTION
**Pull request by stolowski**
Use static attributes stored in conns when reloading connections.

### CODE REVIEW

File: interfaces/connection.go [hunk 1]
```diff
@@ -111,6 +111,16 @@ func NewConnectedSlot(slot *snap.SlotInfo, dynamicAttrs map[string]interface{})
 	}
 }
 
+// NewConnectedSlotReload creates an object representing a connected slot that is reloaded with static attributes
+// restored from state (not from the SlotInfo).
+func NewConnectedSlotReload(slot *snap.SlotInfo, staticAttrs, dynamicAttrs map[string]interface{}) *ConnectedSlot {
+	return &ConnectedSlot{
+		slotInfo:     slot,
+		staticAttrs:  copyAttributes(staticAttrs),
+		dynamicAttrs: utils.NormalizeInterfaceAttributes(dynamicAttrs).(map[string]interface{}),
```
Comment (pedronis) [inline interfaces/connection.go]: the normalize is just for extra safety right?  we store things already normalized in conns?

Comment (stolowski) [reply interfaces/connection.go]: Indeed.

**Comment by stolowski** [comment]
The followup PR https://github.com/snapcore/snapd/pull/5497 implements the patch.

**Comment by stolowski** [comment]
Blocked until #5497 lands.

File: interfaces/repo.go [hunk 2]
```diff
@@ -576,19 +570,54 @@ func (r *Repository) Connect(ref *ConnRef, plugDynamicAttrs, slotDynamicAttrs ma
...
+		return nil, nil, fmt.Errorf(`cannot connect plug "%s:%s" (interface %q) to "%s:%s" (interface %q)`,
 			plugSnapName, plugName, plug.Interface, slotSnapName, slotName, slot.Interface)
 	}
 
+	return plug, slot, nil
+}
+
+// Connect re-establishes a connection between a plug and a slot as part of connection reloading.
+// When connections are reloaded policyCheck is null (we don't check policy again).
+func (r *Repository) ConnectReload(ref *ConnRef, plugStaticAttrs, plugDynamicAttrs, slotStaticAttrs, slotDynamicAttrs map[string]interface{}) (*Connection, error) {
```
Comment (niemeyer) [inline interfaces/repo.go]: I'm missing part of the picture for these API changes. What does it mean to connect-reload? Is it connecting, or is reloading? What does it mean to reload a connection, and why do we fiddle with its static attributes every time we do that?

File: interfaces/repo.go [hunk 3]
```diff
@@ -576,19 +570,54 @@ func (r *Repository) Connect(ref *ConnRef, plugDynamicAttrs, slotDynamicAttrs ma
...
+	if err != nil {
+		return nil, err
+	}
+
+	if _, ok := r.ifaces[plug.Interface]; !ok {
+		return nil, fmt.Errorf("internal error: unknown interface %q", plug.Interface)
+	}
+
+	cplug := NewConnectedPlugReload(plug, plugStaticAttrs, plugDynamicAttrs)
+	cslot := NewConnectedSlotReload(slot, slotStaticAttrs, slotDynamicAttrs)
```
Comment (niemeyer) [inline interfaces/repo.go]: That also seems a bit curious in terms of API. Just reading it out loud seems to make no sense.. It's a new connected plug.. reload?

Seems better to leave the original functions alone and then tuning the results as necessary.

File: interfaces/repo.go (hunk 2 above)
Comment (stolowski) [reply interfaces/repo.go]: Reloading means restoring connections in the repository from "conns" data in the state. We do that in the ifacemgr startup code (InterfaceManager - `reloadConnections` method). In the existing code we just call `repo.Connect(...)` there, but the problem is that it re-creates a connection with static attributes coming from snap yaml data, and dynamic attributes coming from whatever was stored in `conns`. The `ConnectReload` method here allows for re-creating `ConnectedPlug|Slot` which both static and dynamic attributes coming from `conns`.
I agree, the name is probably not very fortunate.

File: interfaces/repo.go (hunk 3 above)
Comment (stolowski) [reply interfaces/repo.go]: I agree, that name is not very fortunate as well...
As for the suggestion about tuning the results - the problem is that ConnectedPlug/Slot types deliberately do not allow modifications of static attributes, therefore this new constructor here.

**Comment by stolowski** [comment]
Changes to NewConnectedPlug/Slot constructors moved to separate PR https://github.com/snapcore/snapd/pull/5807 (it needs to land first).

**Comment by stolowski** [comment]
@niemeyer Simplified to have one NewConnectedPlug/Slot and one Connect method back.

File: interfaces/connection_test.go [hunk 4]
```diff
@@ -329,3 +329,29 @@ func (s *connSuite) TestCopyAttributes(c *C) {
...
+	staticAttrs := map[string]interface{}{
+		"baz": "boom",
+	}
+	dynAttrs := map[string]interface{}{
+		"foo": "bar",
+	}
+	plug := NewConnectedPlug(s.plug, staticAttrs, dynAttrs)
+	c.Assert(plug, NotNil)
+	c.Assert(plug.StaticAttrs(), DeepEquals, staticAttrs)
+	c.Assert(plug.DynamicAttrs(), DeepEquals, dynAttrs)
```
Comment (niemeyer) [inline interfaces/connection_test.go]: Is it supposed to be the exact same value, or a copy?  In either case, I suggest testing the behavior here.

File: interfaces/connection_test.go [hunk 5]
```diff
@@ -329,3 +329,29 @@ func (s *connSuite) TestCopyAttributes(c *C) {
...
+	staticAttrs := map[string]interface{}{
+		"baz": "boom",
+	}
+	dynAttrs := map[string]interface{}{
+		"foo": "bar",
+	}
+	slot := NewConnectedSlot(s.slot, staticAttrs, dynAttrs)
+	c.Assert(slot, NotNil)
+	c.Assert(slot.StaticAttrs(), DeepEquals, staticAttrs)
+	c.Assert(slot.DynamicAttrs(), DeepEquals, dynAttrs)
```
Comment (niemeyer) [inline interfaces/connection_test.go]: Same.

**Pull request merged**